### PR TITLE
feat(showHide): Introduce directives to notify child directives on hi…

### PR DESCRIPTION
**Pull request by kseamon**
…de/show.

Features extreme tail-hooking technology.
Step 2 of addressing #4314 and #4336

### CODE REVIEW

**Comment by kseamon** [comment]
@jelbourn

File: src/components/showHide/showHide.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+/**
+ * @ngdoc module
+ * @name material.components.showHide
+ */
+
+angular.module('material.components.showHide', [
+  'material.core'
+])
+  .directive('ngShow', createDirective('ngShow', true))
```
Comment (jelbourn) [inline src/components/showHide/showHide.js]: I would move part of the later comment here, saying something like
```js
// Create our own directives 'ngShow' and 'ngHide' that are run in 
// _addition_ to the normal ngShow/ngHide. We use these to...
```

**Comment by jelbourn** [comment]
LGTM

File: src/components/showHide/showHide.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+function createDirective(name, targetValue) {
+  return ['$mdUtil', function($mdUtil) {
+    return {
+      restrict: 'A',
+      multiElement: true,
+      link: function($scope, $element, $attr) {
+        $scope.$watch($attr[name], function(value) {
+          if (!!value === targetValue) {
+            $scope.$broadcast('$md-resize');
+            $mdUtil.dom.animator.waitTransitionEnd($element).then(function() {
```
Comment (ThomasBurleson) [inline src/components/showHide/showHide.js]: @kseamon - why are you broadcasting 2x?

File: src/components/showHide/showHide.spec.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      var element = $compile('<div ng-hide="hide"></div>')(scope);
+      scope.$apply();
+      expect(spy).not.toHaveBeenCalled();
+
+      scope.hide = false;
+      scope.$apply();
+      expect(spy).toHaveBeenCalled();
+
+      spy.calls.reset();
+      defered.resolve();
```
Comment (ThomasBurleson) [inline src/components/showHide/showHide.spec.js]: What are you doing here... it is not clear.

File: src/components/showHide/showHide.spec.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      scope.show = false;
+      var element = $compile('<div ng-show="show"></div>')(scope);
+      scope.$apply();
+      expect(spy).not.toHaveBeenCalled();
+
+      scope.show = true;
+      scope.$apply();
+      expect(spy).toHaveBeenCalled();
+
+      spy.calls.reset();
```
Comment (ThomasBurleson) [inline src/components/showHide/showHide.spec.js]: Can you break this group out into a understand function call ?

File: src/components/showHide/showHide.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+// Add additional handlers to ng-show and ng-hide that notify directives
+// contained within that they should recompute their size.
+function createDirective(name, targetValue) {
+  return ['$mdUtil', function($mdUtil) {
+    return {
+      restrict: 'A',
+      multiElement: true,
+      link: function($scope, $element, $attr) {
+        $scope.$watch($attr[name], function(value) {
+          if (!!value === targetValue) {
```
Comment (ThomasBurleson) [inline src/components/showHide/showHide.js]: :+1:

File: src/components/showHide/showHide.js (hunk 2 above)
Comment (kseamon) [inline src/components/showHide/showHide.js]: Yeah, my concern was that not every element has transitions, and 3s is longer than I'd like to wait in those cases.

File: src/components/showHide/showHide.js (hunk 1 above)
Comment (kseamon) [inline src/components/showHide/showHide.js]: Done

File: src/components/showHide/showHide.spec.js (hunk 3 above)
Comment (kseamon) [inline src/components/showHide/showHide.spec.js]: Added comments

File: src/components/showHide/showHide.spec.js (hunk 4 above)
Comment (kseamon) [inline src/components/showHide/showHide.spec.js]: I think the comments added are sufficient. Have a look.

**Pull request closed without merging**